packages/safe-ds-lang/src/language/generation/safe-ds-python-generator.ts

```diff
@@ -775,6 +793,168 @@ export class SafeDsPythonGenerator {
+            },
+        )}], [${joinToNode(hiddenParameters, (param) => param, { separator: ', ' })}])`;
+    }
+
+    private getMemoizedCallHiddenParameters(expression: SdsCall, frame: GenerationInfoFrame): CompositeGeneratorNode[] {
+        const impurityReasons = this.purityComputer.getImpurityReasonsForExpression(expression);
+        const hiddenParameters: CompositeGeneratorNode[] = [];
+        for (const reason of impurityReasons) {
+            if (reason instanceof FileRead) {
+                if (typeof reason.path === 'string') {
+                    hiddenParameters.push(
+                        expandTracedToNode(
+                            expression,
+                        )`${RUNNER_SERVER_PIPELINE_MANAGER_PACKAGE}.runner_filemtime('${reason.path}')`,
+                    );
+                } else if (isSdsParameter(reason.path)) {
+                    const argument = this.nodeMapper
+                        .parametersToArguments([reason.path], expression.argumentList.arguments)
+                        .get(reason.path);
+                    if (!argument) {
+                        /* c8 ignore next 4 */
+                        throw new Error(
+                            'File Read impurity with dependency on parameter is present on call, but no argument has been provided.',
+                        );
+                    }
+                    hiddenParameters.push(
+                        expandTracedToNode(
+                            argument,
+                        )`${RUNNER_SERVER_PIPELINE_MANAGER_PACKAGE}.runner_filemtime(${this.generateArgument(
+                            argument,
+                            frame,
+                        )})`,
+                    );
+                }
+            }
+        }
+        return hiddenParameters;
+    }
+
+    private getFullyQualifiedReferenceName(expression: SdsExpression, declaration: SdsDeclaration): string | undefined {
+        // Root Node is always a module.
+        const currentModule = <SdsModule>findRootNode(expression);
+        const targetModule = <SdsModule>findRootNode(declaration);
+        if (currentModule === targetModule) {
+            return `${
+                this.builtinAnnotations.getPythonModule(targetModule) || currentModule.name
+            }.${this.getPythonNameOrDefault(declaration)}`;
+        }
+        for (const value of getImports(currentModule)) {
+            // Verify same package
+            if (value.package !== targetModule.name) {
+                continue;
+            }
+            if (isSdsQualifiedImport(value)) {
+                const importedDeclarations = getImportedDeclarations(value);
+                for (const importedDeclaration of importedDeclarations) {
+                    if (declaration === importedDeclaration.declaration?.ref) {
+                        return `${
+                            this.builtinAnnotations.getPythonModule(targetModule) || value.package
```

Comment:
Abstract this into its own method, akin to `getPythonNameOrDefault` (`getPythonModuleOrDefault`?).